Revise page about multiple zones (#23498)
kubernetes/website: content/en/docs/setup/best-practices/multiple-zones.md

- justinsb
- quinton-hoole
title: Running in multiple zones
weight: 20
content_type: concept
---

<!-- overview -->

This page describes running Kubernetes across multiple zones.

<!-- body -->

## Background

Kubernetes is designed so that a single Kubernetes cluster can run
across multiple failure zones, typically where these zones fit within
a logical grouping called a _region_. Major cloud providers define a region
as a set of failure zones (also called _availability zones_) that provide
a consistent set of features: within a region, each zone offers the same
APIs and services.

Typical cloud architectures aim to minimize the chance that a failure in
one zone also impairs services in another zone.

## Control plane behavior

> All [control plane components](https://kubernetes.io/docs/concepts/overview/components/#control-plane-components)

Reviewer: nit: use relative link target.

> All [control plane components](https://kubernetes.io/docs/concepts/overview/components/#control-plane-components)

Author: ```suggestion
All [control plane components](/docs/concepts/overview/components/#control-plane-components)
```

support running as a pool of interchangable resources, replicated per
component.

When you deploy a cluster control plane you can place a replica of each
component in each failure zone that you have used. For example, if you
have a choice of 4 failure zones and you choose to use just 3 of these for
your cluster, ensure that there is a scheduler, an API server, a
controller manager and an etcd store in each of the 3 failure zones you
selected. If you are running a cloud controller manager then you should
replicate this across the 3 selected zones as well.

{{< note >}}
Kubernetes does not provide cross-zone resilience for the API server
endpoints. You can use various techniques to improve availability for
the cluster API server, including DNS round-robin, SRV records, or
a third-party load balancing solution with health checking.
{{< /note >}}

## Node behavior

Kubernetes automatically spreads the Pods for a

> {{< glossary_tooltip text="Deployment" term_id="deployment" >}} or

Reviewer 2: nit: Is this supposed to be:
Deployment, ReplicaSet or service?

> {{< glossary_tooltip text="Deployment" term_id="deployment" >}} or
> {{< glossary_tooltip text="ReplicaSet" term_id="replica-set" >}}
> or service across different nodes in a cluster. This spreading helps

Author: ```suggestion
workload resources (such as {{< glossary_tooltip text="Deployment" term_id="deployment" >}}
or {{< glossary_tooltip text="StatefulSet" term_id="statefulset" >}})
across different nodes in a cluster. This spreading helps
```

storage in that class may use.
See [Allowed topologies](/docs/concepts/storage/storage-classes/#allowed-topologies)
for more details.

## Networking

Kubernetes itself does not provide zone-aware networking. You can use a
[network plugin](docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/)
to configure cluster networking, and that network solution might have zone-specific
elements. For example, if your cloud provider supports Services with
`type=LoadBalancer`, the load balancer might only send traffic to Pods running in the
same zone. Check your cloud provider's documentation for details.

For custom or on-premises deployments, similar considerations apply.
{{< glossary_tooltip text="Service" term_id="service" >}} and
{{< glossary_tooltip text="Ingress" term_id="ingress" >}} behavior, including handling
of different failure zones, does vary depending on exactly how your cluster is set up.

## Fault recovery

When you set up your cluster, you might also need to consider whether and how
your setup can restore service if all of the failure zones in a region go
off-line at the same time. For example, do you rely on there being at least

> one running node so that cluster-critical Pods can perform repair work?

Reviewer 2: nit: The question could be rephrased.
For example, do not rely on ...

Author: ```suggestion
Make sure that any cluster-critical repair work does not rely
on there being at least one healthy node in your cluster. For example: if all nodes
are unhealthy, you might need to run a repair Job with a special
{{< glossary_tooltip text="toleration" term_id="toleration" >}} so that the repair
can complete enough to bring at least one node into service.
```

Kubernetes doesn't come with an answer for this challenge; however, it's
something to consider.

## {{% heading "whatsnext" %}}

> If you want to learn more about how the scheduler places Pods in your cluster,

Reviewer 2: nit: Does it make sense to say, "learn more about how the scheduler"? Instead:
To learn how the scheduler distributes [or places] Pods in a cluster, visit [or see] ...

> If you want to learn more about how the scheduler places Pods in your cluster,

Author: ```suggestion
To learn how the scheduler places Pods in a cluster, honoring the configured constraints,
```

- justinsb
- quinton-hoole
title: Running in multiple zones
weight: 20
content_type: concept
---

<!-- overview -->

This page describes running Kubernetes across multiple zones.

<!-- body -->

## Background

Kubernetes is designed so that a single Kubernetes cluster can run
across multiple failure zones, typically where these zones fit within
a logical grouping called a _region_. Major cloud providers define a region
as a set of failure zones (also called _availability zones_) that provide
a consistent set of features: within a region, each zone offers the same
APIs and services.

Typical cloud architectures aim to minimize the chance that a failure in
one zone also impairs services in another zone.

## Control plane behavior

All [control plane components](/docs/concepts/overview/components/#control-plane-components)
support running as a pool of interchangable resources, replicated per
component.

When you deploy a cluster control plane you can place a replica of each

> component in each failure zone that you have used. For example, if you

Reviewer 2: nit: simplify:
When you deploy a cluster control plane, place a replica of each
control plane component in each failure zone.

have a choice of 4 failure zones and you choose to use just 3 of these for
your cluster, ensure that there is a scheduler, an API server, a
controller manager and an etcd store in each of the 3 failure zones you
selected. If you are running a cloud controller manager then you should
replicate this across the 3 selected zones as well.

{{< note >}}
Kubernetes does not provide cross-zone resilience for the API server
endpoints. You can use various techniques to improve availability for
the cluster API server, including DNS round-robin, SRV records, or
a third-party load balancing solution with health checking.
{{< /note >}}

## Node behavior

Kubernetes automatically spreads the Pods for a

> workload resources (such as {{< glossary_tooltip text="Deployment" term_id="deployment" >}}

Reviewer 2: cleanup: remove `a` in `for a workload resources`

or {{< glossary_tooltip text="StatefulSet" term_id="statefulset" >}})
across different nodes in a cluster. This spreading helps
reduce the impact of failures.

When nodes start up, the kubelet on each node automatically adds
{{< glossary_tooltip text="labels" term_id="label" >}} to the the Node object

> that represents that specific kubelet in the Kubernetes API.

Reviewer 2: cleanup: `the the`
possible cleanup: s/Node object/Node/

These labels can include

> [zone information](/docs/reference/kubernetes-api/labels-annotations-taints/#topologykubernetesiozone)

Reviewer 2: nit: Add period at end of sentence

If your cluster spans multiple zones or regions, you can use node labels
in conjunction with
[Pod topology spread constraints](/docs/concepts/workloads/pods/pod-topology-spread-constraints/)
to control how Pods are spread across your cluster among fault domains.

> such as regions, zones, and even specific nodes.

Reviewer 2: Check the sentences (rebase mixup).
`how Pods are spread across your cluster among fault domains.
such as regions, zones, and even specific nodes.`

3 replicas of a StatefulSet are all running in different zones to each
other, whenever that is feasible. You can define this declaratively
without explicitly defining which availability zones are in use for
each workload.

### Distributing nodes across zones

Kubernetes' core does not create nodes for you; you need to do that yourself,
or use a tool such as the [Cluster API](https://cluster-api.sigs.k8s.io/) to
manage nodes on your behalf.

Using tools such as the Cluster API you can define sets of machines to run as
worker nodes for your cluster across multiple failure domains, and rules to
automatically heal the cluster in case of whole-zone service disruption.

## Manual zone assignment for Pods

You can apply [node selector constraints](/docs/concepts/scheduling-eviction/assign-pod-node/#nodeselector)
to Pods that you create, as well as to Pod templates in workload resources
such as Deployment, StatefulSet, or Job.

## Storage access for zones

When persistent volumes are created, the `PersistentVolumeLabel`
[admission controller](/docs/reference/access-authn-authz/admission-controllers/)
automatically adds zone labels to any PersistentVolumes that are linked to a specific

> zone. The {< glossary_tooltip text="scheduler" term_id="kube-scheduler" >}} then ensures,

Reviewer 2: ☝️ This tooltip is not rendering

through its `NoVolumeZoneConflict` predicate, that pods which claim a given PersistentVolume
are only placed into the same zone as that volume.

You can specify a {{< glossary_tooltip text="StorageClass" term_id="storage-class" >}}

> for PersistentVolumeClaims that specifies the failure domains (zones) that the

Reviewer 2: nit wording/break up sentence. Would this work?

To learn about configuring a StorageClass with failure domains or zones, see [Allowed topologies](/docs/concepts/storage/storage-classes/#allowed-topologies).

storage in that class may use.
See [Allowed topologies](/docs/concepts/storage/storage-classes/#allowed-topologies)
for more details.

## Networking

Kubernetes itself does not provide zone-aware networking. You can use a

> [network plugin](docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/)

Reviewer 2: nit: Kubernetes does not provide ...

to configure cluster networking, and that network solution might have zone-specific
elements. For example, if your cloud provider supports Services with
`type=LoadBalancer`, the load balancer might only send traffic to Pods running in the
same zone. Check your cloud provider's documentation for details.

For custom or on-premises deployments, similar considerations apply.
{{< glossary_tooltip text="Service" term_id="service" >}} and
{{< glossary_tooltip text="Ingress" term_id="ingress" >}} behavior, including handling
of different failure zones, does vary depending on exactly how your cluster is set up.

## Fault recovery

When you set up your cluster, you might also need to consider whether and how
your setup can restore service if all of the failure zones in a region go
off-line at the same time. For example, do you rely on there being at least

> Make sure that any cluster-critical repair work does not rely

Reviewer 2: @sftim. I read through the page again. There are a few places to edit.
Have a look at:
`For example, do you rely on there being at least
Make sure that any cluster-critical repair work does not rely`
